src/pymor/vectorarrays/interfaces.py

```diff
@@ -392,7 +392,7 @@ def norm(self, product=None):
         if product is None:
             return self.l2_norm()
         else:
-            return np.sqrt(product.pairwise_apply2(self, self))
+            return np.sqrt(product.pairwise_apply2(self, self).real)
```

Comment:
Do we want to have some error checking here? If there is a non-zero imaginary part (because `product` is non-hermitian) we won't notice. It might be hard to find such an error.

Reply:
Yes, I agree we should be careful here.
I'm not sure what would be a good threshold for raising an exception. Maybe only raise a warning whenever the imaginary part is non-zero?
Or, does it make sense to define the inner product for non-hermitian `product` as the inner product using the hermitian part of `product`? That would then be exactly the real part of `product.pairwise_apply2(self, self)`.

Reply:
> I'm not sure what would be a good threshold for raising an exception. Maybe only raise a warning whenever the imaginary part is non-zero?

I guess we would also need to choose a threshold even when we only issue a warning. Probably, we could make this a default like [here](https://github.com/pymor/pymor/blob/master/src/pymor/operators/constructions.py#L1050). No idea what a reasonable default tolerance would be or if it should be relative to the real part.

> Or, does it make sense to define the inner product for non-hermitian product as the inner product using the hermitian part of product?

Maybe it can make sense in some situations, but usually there will be an error somewhere ..

Reply:
I guess it makes sense to have a similar logic to `constructions.induced_norm`. There is an [error bound for inner product of two vectors](https://doi.org/10.1137/120894488) which scales linearly with the dimension, but it also depends on the norms of the vectors. I would be fine with raising an exception when `np.abs(product.pairwise_apply2(self, self).imag) > tol`, where `tol` is a default like in `induced_norm` (and `raise_negative` is appropriately renamed).